File: Day226/leafUnderBudget.go
package main

import (
	"fmt"
	"sort"
)

type Node struct {
	data  int
	left  *Node
	right *Node
}

type Solution struct {
	count int
	list  []int
}

func (s *Solution) getCount(node *Node, bud int) int {
	s.list = make([]int, 0)
	s.inorder(node)
	sort.Ints(s.list)
	answer := 0
	sum := 0
	for i := 0; i < len(s.list); i++ {
		if sum+s.list[i] <= bud {
			sum += s.list[i]
			answer++
		}
	}
	return answer
}

func (s *Solution) inorder(root *Node) {
	if root == nil {
		return
	}
	s.count++
	s.inorder(root.left)
	if root.left == nil && root.right == nil {
		s.list = append(s.list, s.count)
	}
	s.inorder(root.right)
	s.count--
}

File: Day226/leafUnderBudget.java
class Solution{
    int count = 0;
    List<Integer> list;
    public int getCount(Node node, int bud)
    {
        //code here  
        list = new ArrayList<>();
        inorder(node);
        Collections.sort(list);
        int answer = 0;
        int sum = 0;
        for (int i=0; i<list.size(); i++) {
            if (sum+list.get(i)<=bud) {
                sum+= list.get(i);
                answer++;
            }
        }
        // System.out.println(list);
        return answer;
    }
    
    public void inorder(Node root) {
        if (root == null) {
            return;
        }
        count+=1;
        inorder(root.left);
        if (root.left == null && root.right == null) {
            list.add(count);
        }
        inorder(root.right);
        count--;
        
    }
}